docs/source/format/CDeviceDataInterface.rst

```diff
@@ -650,6 +652,331 @@ The stream source is not assumed to be thread-safe. Consumers wanting to
+  callback on the async stream handler.
+
+.. c:member:: void* ArrowAsyncProducer.private_data
+
+  *Optional.* An opaque pointer to producer-provided specific data.
+
+  Consumers *MUST NOT* process this member, the lifetime is owned by the producer
+  that constructed this object.
+
+Error Handling
+''''''''''''''
+
+Unlike the regular C Stream interface, the Async interface allows for errors to flow in
+both directions. As a result, error handling can be slightly more complex. Thus this spec
+designates the following rules:
+
+* If the producer encounters an error during processing, it should call the ``on_error``
+callback, and then call ``release`` after it returns.
+
+* If ``on_schema`` or ``on_next_task`` returns a non-zero integer value, the producer *should not*
+call the ``on_error`` callback, but instead should eventually call ``release`` at some point
+before or after any logging or processing of the error code.
+
+Result lifetimes
+''''''''''''''''
+
+The ``ArrowSchema`` passed to the ``on_schema`` callback must be released independantly,
+with the object itself needing to be moved to a consumer owned ``ArrowSchema`` object. The
+``ArrowSchema*`` passed as a parameter to the callback *MUST NOT* be stored and kept.
+
+The ``ArrowAsyncTask`` object provided to ``on_next_task`` is owned by the producer and
+will be cleaned up during the invocation of calling ``get_data`` on it. If the consumer
+doesn't care about the data, it should pass ``NULL`` instead of a valid ``ArrowDeviceArray*``.
+
+The ``const char*`` error ``message`` and ``metadata`` which are passed to ``on_error``
+are only valid within the scope of the ``on_error`` function itself. They must be copied
+if it is necessary for them to exist after it returns.
+
+Stream Handler Lifetime
+'''''''''''''''''''''''
+
+Lifetime of the async stream handler is managed using a release callback with similar
+usage as in :ref:`C data interface <c-data-interface-released>`.
+
+ArrowAsyncProducer Lifetime
+'''''''''''''''''''''''''''
+
+The lifetime of the ``ArrowAsyncProducer`` passed to ``on_schema`` is owned by the producer
+itself and should be managed by it. It *MUST* remain valid at least until just before
+calling ``release`` on the stream handler object.
+
+Thread safety
+'''''''''''''
+
+All handler functions should only be called in a serialized manner, but are not guaranteed
+to be called from the same thread every time. A producer should wait for handler callbacks to
+return before calling the next handler callback, and before calling the ``release`` callback.
+As a result, back-pressure is managed by how long it takes for the ``on_next`` handler to return.
+
+The ``ArrowAsyncDeviceStreamHandler`` object should be able to handle callbacks as soon as
+it is passed to the producer, any initialization should be performed before it is provided.
+
+Possible Sequence Diagram
+-------------------------
+
+.. mermaid::
```

Comment:
Is there an easy way to see a rendered version of this? :-)

Reply:
You can use https://mermaid.live/edit to see a rendered version. You can also follow [this link](https://mermaid.live/edit#pako:eNqdU01vwjAM_StRToXR_oAeKqFx2GXaJHashExiICJNunxsQ4j_PqddKRsbh_nktM_Pzy_OkQsrkZfc42tEI3ChYOugqQ2juLfGxwZdXlV3z87KKNCVbO6cfZ_7gxELfFMCl8EhNA9gpEY37SsHdJ5KB5qSWbPyYocNZCPJAJ3OeuZlB5hOriT8puDcp3JpAB8yQ4V9aQuuT1Joa1sWWxYsMyyoBv3476fgb3oNfoRVAL-_kPxCx0FgCjSyP5AHbGMdQxA7lopGzHmQPyZJnMUWw0pCgKwoii_-jrtP1-Tznj21QVkDWh9-Ia-q_LZNAuiSdXZN3hnUWu_VWiNzdAXKKLMde_zboYsmo44qHzluLlS6WY3gMZvwGSc8CZO0sMfEWPNAy4I1LymVuIGoQ81rcyIoxGCXxMnL4CLOeGzJ2WG_h48oVbDusX8D3VM4fQI_2wk4) which should already have the diagram in it to view :)